cmd/crossplane/main.go

```diff
@@ -49,6 +50,13 @@ func main() {
 		ctrl.SetLogger(zl)
 	}
 
+	provider, err := metrics.NewMeterProvider()
+	kingpin.FatalIfError(err, "cannot create metrics provider")
+	kingpin.FatalIfError(metrics.Instrument(provider), "cannot instrument metrics packages")
+	server := metrics.NewServer(logging.NewLogrLogger(zl.WithName("metrics")))
+	server.Start()
+	defer server.Stop()
```

Comment:
The controller manager actually already starts a metrics server for us on port `8080`, so we are currently serving metrics, just not collecting them or documenting how to do so. I think we should use this functionality and extend it to include other metrics we find relevant. Helpful links
- [Kubebuilder Book -  Exporting Metrics  for Prometheus](https://book.kubebuilder.io/reference/metrics.html)
- [AddMetricsExtraHandler](https://github.com/kubernetes-sigs/controller-runtime/blob/8d0107636985eeedeb8c3c666abda5f602aa92be/pkg/manager/manager.go#L69)

Reply:
Oh interesting, I wasn't aware of this, I'm gonna look into it. Now I'm wondering if we need our instrumentation or rather just tap into this and potentially just add extra crossplane related metrics to id.

Reply:
Yeah, I'd definitely tap into the existing metrics. I'd prefer to be using OpenTelemetry, but probably being consistent and integrated with controller-runtime is the path of least resistance. It does seem like there's some interest in moving controller-runtime to OpenTelemetry per https://github.com/kubernetes-sigs/controller-runtime/issues/305.

Reply:
I am assigned on that upstream issue :sweat_smile: hopefully I can get that implemented quickly for us to use here :+1:

Reply:
Awesome! So in that case the change in this PR can reduce essentially to:

```diff
--- a/cluster/charts/crossplane/templates/deployment.yaml
+++ b/cluster/charts/crossplane/templates/deployment.yaml
@@ -42,6 +42,8 @@ spec:
         resources:
           {{- toYaml .Values.resourcesCrossplane | nindent 12 }}
         ports:
+        - name: http
+          containerPort: 8080
```

That said, this doesn't still incorporate any additional crossplane specific metrics. Is there some that can be immediately thought of?

<details>
   <summary>This is the current exposed metrics for crossplane core (click to expand).</summary>

```
# HELP go_gc_duration_seconds A summary of the GC invocation durations.
# TYPE go_gc_duration_seconds summary
go_gc_duration_seconds{quantile="0"} 4.0587e-05
go_gc_duration_seconds{quantile="0.25"} 0.00011317
go_gc_duration_seconds{quantile="0.5"} 0.000134173
go_gc_duration_seconds{quantile="0.75"} 0.00020736
go_gc_duration_seconds{quantile="1"} 0.000355866
go_gc_duration_seconds_sum 0.000851156
go_gc_duration_seconds_count 5
# HELP go_goroutines Number of goroutines that currently exist.
# TYPE go_goroutines gauge
go_goroutines 109
# HELP go_info Information about the Go environment.
# TYPE go_info gauge
go_info{version="go1.15.2"} 1
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 6.762912e+06
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 1.2796024e+07
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 1.448036e+06
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 58043
# HELP go_memstats_gc_cpu_fraction The fraction of this program's available CPU time used by the GC since the program started.
# TYPE go_memstats_gc_cpu_fraction gauge
go_memstats_gc_cpu_fraction 0.039225983442468006
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 5.373816e+06
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 6.762912e+06
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 5.5853056e+07
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 9.814016e+06
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 40702
# HELP go_memstats_heap_released_bytes Number of heap bytes released to OS.
# TYPE go_memstats_heap_released_bytes gauge
go_memstats_heap_released_bytes 5.4951936e+07
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 6.5667072e+07
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 1.6050463810900233e+09
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 0
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 98745
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 20832
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 32768
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 202232
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 212992
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 1.1632768e+07
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 2.190372e+06
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 1.441792e+06
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 1.441792e+06
# HELP go_memstats_sys_bytes Number of bytes obtained from system.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 7.6366848e+07
# HELP go_threads Number of OS threads created.
# TYPE go_threads gauge
go_threads 17
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 0.37
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 1.048576e+06
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 10
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 3.8326272e+07
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.60504637745e+09
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 7.61315328e+08
# HELP process_virtual_memory_max_bytes Maximum amount of virtual memory available in bytes.
# TYPE process_virtual_memory_max_bytes gauge
process_virtual_memory_max_bytes -1
# HELP rest_client_request_latency_seconds Request latency in seconds. Broken down by verb and URL.
# TYPE rest_client_request_latency_seconds histogram
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="GET",le="0.001"} 3
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="GET",le="0.002"} 9
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="GET",le="0.004"} 23
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="GET",le="0.008"} 27
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="GET",le="0.016"} 27
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="GET",le="0.032"} 27
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="GET",le="0.064"} 27
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="GET",le="0.128"} 59
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="GET",le="0.256"} 62
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="GET",le="0.512"} 65
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="GET",le="+Inf"} 65
rest_client_request_latency_seconds_sum{url="https://10.96.0.1:443/%7Bprefix%7D",verb="GET"} 4.9588399830000025
rest_client_request_latency_seconds_count{url="https://10.96.0.1:443/%7Bprefix%7D",verb="GET"} 65
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="POST",le="0.001"} 0
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="POST",le="0.002"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="POST",le="0.004"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="POST",le="0.008"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="POST",le="0.016"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="POST",le="0.032"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="POST",le="0.064"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="POST",le="0.128"} 2
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="POST",le="0.256"} 2
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="POST",le="0.512"} 2
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="POST",le="+Inf"} 2
rest_client_request_latency_seconds_sum{url="https://10.96.0.1:443/%7Bprefix%7D",verb="POST"} 0.090575015
rest_client_request_latency_seconds_count{url="https://10.96.0.1:443/%7Bprefix%7D",verb="POST"} 2
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="PUT",le="0.001"} 0
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="PUT",le="0.002"} 3
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="PUT",le="0.004"} 4
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="PUT",le="0.008"} 20
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="PUT",le="0.016"} 20
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="PUT",le="0.032"} 20
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="PUT",le="0.064"} 20
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="PUT",le="0.128"} 21
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="PUT",le="0.256"} 21
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="PUT",le="0.512"} 21
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/%7Bprefix%7D",verb="PUT",le="+Inf"} 21
rest_client_request_latency_seconds_sum{url="https://10.96.0.1:443/%7Bprefix%7D",verb="PUT"} 0.18058522299999993
rest_client_request_latency_seconds_count{url="https://10.96.0.1:443/%7Bprefix%7D",verb="PUT"} 21
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/api?timeout=32s",verb="GET",le="0.001"} 0
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/api?timeout=32s",verb="GET",le="0.002"} 0
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/api?timeout=32s",verb="GET",le="0.004"} 0
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/api?timeout=32s",verb="GET",le="0.008"} 0
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/api?timeout=32s",verb="GET",le="0.016"} 0
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/api?timeout=32s",verb="GET",le="0.032"} 0
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/api?timeout=32s",verb="GET",le="0.064"} 0
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/api?timeout=32s",verb="GET",le="0.128"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/api?timeout=32s",verb="GET",le="0.256"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/api?timeout=32s",verb="GET",le="0.512"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/api?timeout=32s",verb="GET",le="+Inf"} 1
rest_client_request_latency_seconds_sum{url="https://10.96.0.1:443/api?timeout=32s",verb="GET"} 0.094454191
rest_client_request_latency_seconds_count{url="https://10.96.0.1:443/api?timeout=32s",verb="GET"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/apis?timeout=32s",verb="GET",le="0.001"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/apis?timeout=32s",verb="GET",le="0.002"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/apis?timeout=32s",verb="GET",le="0.004"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/apis?timeout=32s",verb="GET",le="0.008"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/apis?timeout=32s",verb="GET",le="0.016"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/apis?timeout=32s",verb="GET",le="0.032"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/apis?timeout=32s",verb="GET",le="0.064"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/apis?timeout=32s",verb="GET",le="0.128"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/apis?timeout=32s",verb="GET",le="0.256"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/apis?timeout=32s",verb="GET",le="0.512"} 1
rest_client_request_latency_seconds_bucket{url="https://10.96.0.1:443/apis?timeout=32s",verb="GET",le="+Inf"} 1
rest_client_request_latency_seconds_sum{url="https://10.96.0.1:443/apis?timeout=32s",verb="GET"} 0.000602968
rest_client_request_latency_seconds_count{url="https://10.96.0.1:443/apis?timeout=32s",verb="GET"} 1
# HELP rest_client_requests_total Number of HTTP requests, partitioned by status code, method, and host.
# TYPE rest_client_requests_total counter
rest_client_requests_total{code="200",host="10.96.0.1:443",method="GET"} 72
rest_client_requests_total{code="200",host="10.96.0.1:443",method="PUT"} 21
rest_client_requests_total{code="201",host="10.96.0.1:443",method="POST"} 2
rest_client_requests_total{code="404",host="10.96.0.1:443",method="GET"} 1
# HELP workqueue_adds_total Total number of adds handled by workqueue
# TYPE workqueue_adds_total counter
workqueue_adds_total{name="defined/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_adds_total{name="offered/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_adds_total{name="packages/configuration.pkg.crossplane.io"} 0
workqueue_adds_total{name="packages/configurationrevision.pkg.crossplane.io"} 0
workqueue_adds_total{name="packages/provider.pkg.crossplane.io"} 0
workqueue_adds_total{name="packages/providerrevision.pkg.crossplane.io"} 0
# HELP workqueue_depth Current depth of workqueue
# TYPE workqueue_depth gauge
workqueue_depth{name="defined/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_depth{name="offered/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_depth{name="packages/configuration.pkg.crossplane.io"} 0
workqueue_depth{name="packages/configurationrevision.pkg.crossplane.io"} 0
workqueue_depth{name="packages/provider.pkg.crossplane.io"} 0
workqueue_depth{name="packages/providerrevision.pkg.crossplane.io"} 0
# HELP workqueue_longest_running_processor_seconds How many seconds has the longest running processor for workqueue been running.
# TYPE workqueue_longest_running_processor_seconds gauge
workqueue_longest_running_processor_seconds{name="defined/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_longest_running_processor_seconds{name="offered/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_longest_running_processor_seconds{name="packages/configuration.pkg.crossplane.io"} 0
workqueue_longest_running_processor_seconds{name="packages/configurationrevision.pkg.crossplane.io"} 0
workqueue_longest_running_processor_seconds{name="packages/provider.pkg.crossplane.io"} 0
workqueue_longest_running_processor_seconds{name="packages/providerrevision.pkg.crossplane.io"} 0
# HELP workqueue_queue_duration_seconds How long in seconds an item stays in workqueue before being requested
# TYPE workqueue_queue_duration_seconds histogram
workqueue_queue_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="1e-08"} 0
workqueue_queue_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="1e-07"} 0
workqueue_queue_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="1e-06"} 0
workqueue_queue_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="9.999999999999999e-06"} 0
workqueue_queue_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="9.999999999999999e-05"} 0
workqueue_queue_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="0.001"} 0
workqueue_queue_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="0.01"} 0
workqueue_queue_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="0.1"} 0
workqueue_queue_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="1"} 0
workqueue_queue_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="10"} 0
workqueue_queue_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="+Inf"} 0
workqueue_queue_duration_seconds_sum{name="defined/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_queue_duration_seconds_count{name="defined/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_queue_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="1e-08"} 0
workqueue_queue_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="1e-07"} 0
workqueue_queue_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="1e-06"} 0
workqueue_queue_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="9.999999999999999e-06"} 0
workqueue_queue_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="9.999999999999999e-05"} 0
workqueue_queue_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="0.001"} 0
workqueue_queue_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="0.01"} 0
workqueue_queue_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="0.1"} 0
workqueue_queue_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="1"} 0
workqueue_queue_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="10"} 0
workqueue_queue_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="+Inf"} 0
workqueue_queue_duration_seconds_sum{name="offered/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_queue_duration_seconds_count{name="offered/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="1e-08"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="1e-07"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="1e-06"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="9.999999999999999e-06"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="9.999999999999999e-05"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="0.001"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="0.01"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="0.1"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="1"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="10"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="+Inf"} 0
workqueue_queue_duration_seconds_sum{name="packages/configuration.pkg.crossplane.io"} 0
workqueue_queue_duration_seconds_count{name="packages/configuration.pkg.crossplane.io"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="1e-08"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="1e-07"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="1e-06"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="9.999999999999999e-06"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="9.999999999999999e-05"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="0.001"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="0.01"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="0.1"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="1"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="10"} 0
workqueue_queue_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="+Inf"} 0
workqueue_queue_duration_seconds_sum{name="packages/configurationrevision.pkg.crossplane.io"} 0
workqueue_queue_duration_seconds_count{name="packages/configurationrevision.pkg.crossplane.io"} 0
workqueue_queue_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="1e-08"} 0
workqueue_queue_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="1e-07"} 0
workqueue_queue_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="1e-06"} 0
workqueue_queue_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="9.999999999999999e-06"} 0
workqueue_queue_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="9.999999999999999e-05"} 0
workqueue_queue_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="0.001"} 0
workqueue_queue_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="0.01"} 0
workqueue_queue_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="0.1"} 0
workqueue_queue_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="1"} 0
workqueue_queue_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="10"} 0
workqueue_queue_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="+Inf"} 0
workqueue_queue_duration_seconds_sum{name="packages/provider.pkg.crossplane.io"} 0
workqueue_queue_duration_seconds_count{name="packages/provider.pkg.crossplane.io"} 0
workqueue_queue_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="1e-08"} 0
workqueue_queue_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="1e-07"} 0
workqueue_queue_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="1e-06"} 0
workqueue_queue_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="9.999999999999999e-06"} 0
workqueue_queue_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="9.999999999999999e-05"} 0
workqueue_queue_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="0.001"} 0
workqueue_queue_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="0.01"} 0
workqueue_queue_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="0.1"} 0
workqueue_queue_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="1"} 0
workqueue_queue_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="10"} 0
workqueue_queue_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="+Inf"} 0
workqueue_queue_duration_seconds_sum{name="packages/providerrevision.pkg.crossplane.io"} 0
workqueue_queue_duration_seconds_count{name="packages/providerrevision.pkg.crossplane.io"} 0
# HELP workqueue_retries_total Total number of retries handled by workqueue
# TYPE workqueue_retries_total counter
workqueue_retries_total{name="defined/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_retries_total{name="offered/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_retries_total{name="packages/configuration.pkg.crossplane.io"} 0
workqueue_retries_total{name="packages/configurationrevision.pkg.crossplane.io"} 0
workqueue_retries_total{name="packages/provider.pkg.crossplane.io"} 0
workqueue_retries_total{name="packages/providerrevision.pkg.crossplane.io"} 0
# HELP workqueue_unfinished_work_seconds How many seconds of work has been done that is in progress and hasn't been observed by work_duration. Large values indicate stuck threads. One can deduce the number of stuck threads by observing the rate at which this increases.
# TYPE workqueue_unfinished_work_seconds gauge
workqueue_unfinished_work_seconds{name="defined/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_unfinished_work_seconds{name="offered/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_unfinished_work_seconds{name="packages/configuration.pkg.crossplane.io"} 0
workqueue_unfinished_work_seconds{name="packages/configurationrevision.pkg.crossplane.io"} 0
workqueue_unfinished_work_seconds{name="packages/provider.pkg.crossplane.io"} 0
workqueue_unfinished_work_seconds{name="packages/providerrevision.pkg.crossplane.io"} 0
# HELP workqueue_work_duration_seconds How long in seconds processing an item from workqueue takes.
# TYPE workqueue_work_duration_seconds histogram
workqueue_work_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="1e-08"} 0
workqueue_work_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="1e-07"} 0
workqueue_work_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="1e-06"} 0
workqueue_work_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="9.999999999999999e-06"} 0
workqueue_work_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="9.999999999999999e-05"} 0
workqueue_work_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="0.001"} 0
workqueue_work_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="0.01"} 0
workqueue_work_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="0.1"} 0
workqueue_work_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="1"} 0
workqueue_work_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="10"} 0
workqueue_work_duration_seconds_bucket{name="defined/compositeresourcedefinition.apiextensions.crossplane.io",le="+Inf"} 0
workqueue_work_duration_seconds_sum{name="defined/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_work_duration_seconds_count{name="defined/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_work_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="1e-08"} 0
workqueue_work_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="1e-07"} 0
workqueue_work_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="1e-06"} 0
workqueue_work_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="9.999999999999999e-06"} 0
workqueue_work_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="9.999999999999999e-05"} 0
workqueue_work_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="0.001"} 0
workqueue_work_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="0.01"} 0
workqueue_work_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="0.1"} 0
workqueue_work_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="1"} 0
workqueue_work_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="10"} 0
workqueue_work_duration_seconds_bucket{name="offered/compositeresourcedefinition.apiextensions.crossplane.io",le="+Inf"} 0
workqueue_work_duration_seconds_sum{name="offered/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_work_duration_seconds_count{name="offered/compositeresourcedefinition.apiextensions.crossplane.io"} 0
workqueue_work_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="1e-08"} 0
workqueue_work_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="1e-07"} 0
workqueue_work_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="1e-06"} 0
workqueue_work_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="9.999999999999999e-06"} 0
workqueue_work_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="9.999999999999999e-05"} 0
workqueue_work_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="0.001"} 0
workqueue_work_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="0.01"} 0
workqueue_work_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="0.1"} 0
workqueue_work_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="1"} 0
workqueue_work_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="10"} 0
workqueue_work_duration_seconds_bucket{name="packages/configuration.pkg.crossplane.io",le="+Inf"} 0
workqueue_work_duration_seconds_sum{name="packages/configuration.pkg.crossplane.io"} 0
workqueue_work_duration_seconds_count{name="packages/configuration.pkg.crossplane.io"} 0
workqueue_work_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="1e-08"} 0
workqueue_work_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="1e-07"} 0
workqueue_work_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="1e-06"} 0
workqueue_work_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="9.999999999999999e-06"} 0
workqueue_work_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="9.999999999999999e-05"} 0
workqueue_work_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="0.001"} 0
workqueue_work_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="0.01"} 0
workqueue_work_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="0.1"} 0
workqueue_work_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="1"} 0
workqueue_work_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="10"} 0
workqueue_work_duration_seconds_bucket{name="packages/configurationrevision.pkg.crossplane.io",le="+Inf"} 0
workqueue_work_duration_seconds_sum{name="packages/configurationrevision.pkg.crossplane.io"} 0
workqueue_work_duration_seconds_count{name="packages/configurationrevision.pkg.crossplane.io"} 0
workqueue_work_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="1e-08"} 0
workqueue_work_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="1e-07"} 0
workqueue_work_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="1e-06"} 0
workqueue_work_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="9.999999999999999e-06"} 0
workqueue_work_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="9.999999999999999e-05"} 0
workqueue_work_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="0.001"} 0
workqueue_work_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="0.01"} 0
workqueue_work_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="0.1"} 0
workqueue_work_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="1"} 0
workqueue_work_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="10"} 0
workqueue_work_duration_seconds_bucket{name="packages/provider.pkg.crossplane.io",le="+Inf"} 0
workqueue_work_duration_seconds_sum{name="packages/provider.pkg.crossplane.io"} 0
workqueue_work_duration_seconds_count{name="packages/provider.pkg.crossplane.io"} 0
workqueue_work_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="1e-08"} 0
workqueue_work_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="1e-07"} 0
workqueue_work_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="1e-06"} 0
workqueue_work_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="9.999999999999999e-06"} 0
workqueue_work_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="9.999999999999999e-05"} 0
workqueue_work_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="0.001"} 0
workqueue_work_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="0.01"} 0
workqueue_work_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="0.1"} 0
workqueue_work_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="1"} 0
workqueue_work_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="10"} 0
workqueue_work_duration_seconds_bucket{name="packages/providerrevision.pkg.crossplane.io",le="+Inf"} 0
workqueue_work_duration_seconds_sum{name="packages/providerrevision.pkg.crossplane.io"} 0
workqueue_work_duration_seconds_count{name="packages/providerrevision.pkg.crossplane.io"} 0
```
</details>

Reply:
I think just starting with the defaults is fine :+1:

Reply:
I removed the custom metrics server part, and adjusted the ports of deployment. Also note that I added scraping annotations to deployments for ease of use of consumers.